---
title: "Publications"
date: "2016-05-05T21:48:51-07:00"
menu: main
name: "Publications"
url: "/publications/"
weight: 2
---

Follow these links for a comprehensive list of publications, coauthors, and
other information:

- [ORCID](https://orcid.org/0000-0002-2843-6370)
- [NCBI Bibliography](https://www.ncbi.nlm.nih.gov/myncbi/kamil.slowikowski.1/bibliography/public/)
- [Google Scholar](https://scholar.google.com/citations?user=kMP4830AAAAJ)
- [Elsevier Scopus](https://www.scopus.com/authid/detail.uri?authorId=55644101100)

Selected publications are listed below.

<div id="5" class="bt b--black-10 pt5 mt5">

<h1 class="mt0 f3">CUX1 and IκBζ (NFKBIZ) mediate the synergistic inflammatory response to TNF and IL-17A in stromal fibroblasts</h1>

<p><b>Kamil Slowikowski</b>&#42;, Hung N. Nguyen&#42;, Erika H. Noss, Daimon P. Simmons,  Fumitaka Mizoguchi, Gerald F. M. Watts, Michael F. Gurish, Michael B. Brenner, and Soumya Raychaudhuri</p>

<p><i>PNAS</i>, 2020. DOI: <a rel="noopener" target="_blank" href="https://doi.org/10.1073/pnas.1912702117">10.1073/pnas.1912702117</a></p>

<p>
<a class="f5 fw5 link br-pill ba b--black-10 hvr-shadow ph3 pv2 mb2 dib near-black" rel="noopener" target="_blank" href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE129488"><i class="fas fa-box-open"></i> Data</a>
<a class="f5 fw5 link br-pill ba b--black-10 hvr-shadow ph3 pv2 mb2 dib near-black" rel="noopener" target="_blank" href="https://github.com/slowkow/fibrotime"><i class="fab fa-github"></i> GitHub</a>
<a class="f5 fw5 link br-pill hvr-shadow ba b--black-10 ph3 pv2 mb2 dib near-black" rel="noopener" target="_blank" href="https://www.biorxiv.org/content/10.1101/571315v1.full.pdf"><i class="far fa-file-pdf"></i> PDF</a>
<a class="f5 fw5 link br-pill hvr-shadow ba b--black-10 ph3 pv2 mb2 dib near-black" rel="noopener" target="_blank" href="https://www.ncbi.nlm.nih.gov/pubmed/32079724"><i class="fas fa-book-open"></i> PubMed</a>
</p>

<div class="figure mw6 center"><img src="/images/slowikowski2020.png" alt="Figure 2" data-zoomable></img></div>

<h2>Abstract</h2>

<p>The role of stromal fibroblasts in chronic inflammation is unfolding. In
rheumatoid arthritis, leukocyte-derived cytokines TNF and IL-17A work together,
activating fibroblasts to become a dominant source of the hallmark cytokine
IL-6. However, IL-17A alone has minimal effect on fibroblasts. To identify key
mediators of the synergistic response to TNF and IL-17A in human synovial
fibroblasts, we performed time series, dose-response, and gene-silencing
transcriptomics experiments. Here we show that in combination with TNF, IL-17A
selectively induces a specific set of genes mediated by factors including
cut-like homeobox 1 (CUX1) and IκBζ (NFKBIZ). In the promoters of <i>CXCL1</i>, <i>CXCL2</i>,
and <i>CXCL3</i>, we found a putative CUX1-NF-κB binding motif not found elsewhere in
the genome. CUX1 and NF-κB p65 mediate transcription of these genes independent
of LIFR, STAT3, STAT4, and ELF3. Transcription of <i>NFKBIZ</i>, encoding the atypical
IκB factor IκBζ, is IL-17A dose-dependent, and IκBζ only mediates the
transcriptional response to TNF and IL-17A, but not to TNF alone. In
fibroblasts, IL-17A response depends on CUX1 and IκBζ to engage the NF-κB
complex to produce chemoattractants for neutrophil and monocyte
recruitment.</p>

<p class="db center tc">
<a class="f5 fw5 link br-pill ba b--black-10 hvr-shadow ph3 pv2 mb2 dib near-black" rel="noopener" target="_blank" href="https://immunogenomics.io/fibrotime"><i class="fas fa-rocket"></i> View the data</a>
</p>
<div class="db center tc w-70 figure" style="margin-top:2rem">
<video src="/projects/fibrotime-1-fast-m.mp4" style="width:100%;" type="video/mp4" muted="" autoplay="" loop=""></video>
</div>
</div>

<!--<div class="center"><img src="/images/pdf-thumbs/slowikowski2018-montage.jpg" alt="Montage"></img></div>-->

</div>

<div id="4" class="bt b--black-10 pt5 mt5">

<h1 class="mt0 f3">Defining inflammatory cell states in rheumatoid arthritis joint synovial tissues by integrating single-cell transcriptomics and mass cytometry</h1>

<p>Fan Zhang&#42;, Kevin Wei&#42;, <b>Kamil Slowikowski</b>&#42;, Chamith Y. Fonseka&#42;, Deepak A. Rao&#42;, Stephen Kelly, Susan M. Goodman, Darren Tabechian, Laura B. Hughes, Karen Salomon-Escoto, Gerald F. M. Watts, A. Helena Jonsson, Javier Rangel-Moreno, Nida Meednu, Cristina Rozo, William Apruzzese, Thomas M. Eisenhaure, David J. Lieb, David L. Boyle, Arthur M. Mandelin II, Accelerating Medicines Partnership Rheumatoid Arthritis and Systemic Lupus Erythematosus (AMP RA/SLE) Consortium, Brendan F. Boyce, Edward DiCarlo, Ellen M. Gravallese, Peter K. Gregersen, Larry Moreland, Gary S. Firestein, Nir Hacohen, Chad Nusbaum, James A. Lederer, Harris Perlman, Costantino Pitzalis, Andrew Filer, V. Michael Holers, Vivian P. Bykerk, Laura T. Donlin, Jennifer H. Anolik, Michael B. Brenner & Soumya Raychaudhuri</p>

<p><i>Nature Immunology</i>, 2019. DOI: <a rel="noopener" target="_blank" href="https://doi.org/10.1038/s41590-019-0378-1">10.1038/s41590-019-0378-1</a></p>

<p>
<a class="f5 fw5 link br-pill ba b--black-10 hvr-shadow ph3 pv2 mb2 dib near-black" rel="noopener" target="_blank" href="https://www.immport.org/shared/study/SDY998"><i class="fas fa-box-open"></i> Data</a>
<a class="f5 fw5 link br-pill ba b--black-10 hvr-shadow ph3 pv2 mb2 dib near-black" rel="noopener" target="_blank" href="https://github.com/immunogenomics/amp_phase1_ra"><i class="fab fa-github"></i> GitHub</a>
<a class="f5 fw5 link br-pill hvr-shadow ba b--black-10 ph3 pv2 mb2 dib near-black" rel="noopener" target="_blank" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6602051/pdf/nihms-1524330.pdf"><i class="far fa-file-pdf"></i> PDF</a>
<a class="f5 fw5 link br-pill hvr-shadow ba b--black-10 ph3 pv2 mb2 dib near-black" rel="noopener" target="_blank" href="https://www.ncbi.nlm.nih.gov/pubmed/31061532"><i class="fas fa-book-open"></i> PubMed</a>
</p>

<div class="figure mw6 center"><img src="/images/zhang2019.png" alt="Figure 2" data-zoomable></img></div>

<h2>Abstract</h2>

<p>To define the cell populations that drive joint inflammation in rheumatoid
arthritis (RA), we applied single-cell RNA sequencing (scRNA-seq), mass
cytometry, bulk RNA sequencing (RNA-seq) and flow cytometry to T cells, B
cells, monocytes, and fibroblasts from 51 samples of synovial tissue from
patients with RA or osteoarthritis (OA). Utilizing an integrated strategy based
on canonical correlation analysis of 5,265 scRNA-seq profiles, we identified 18
unique cell populations. Combining mass cytometry and transcriptomics revealed
cell states expanded in RA synovia: THY1(CD90)+HLA-DRAhi sublining fibroblasts,
IL1B+ pro-inflammatory monocytes, ITGAX+TBX21+ autoimmune-associated B cells
and PDCD1+ peripheral helper T (TPH) cells and follicular helper T (TFH) cells.
We defined distinct subsets of CD8+ T cells characterized by GZMK+, GZMB+, and
GNLY+ phenotypes. We mapped inflammatory mediators to their source cell
populations; for example, we attributed IL6 expression to THY1+HLA-DRAhi
fibroblasts and IL1B production to pro-inflammatory monocytes. These
populations are potentially key mediators of RA pathogenesis.</p>

<!--<div class="center"><img src="/images/pdf-thumbs/slowikowski2018-montage.jpg" alt="Montage"></img></div>-->

</div>

<div id="3" class="bt b--black-10 pt5 mt5">

<h1 class="mt0 f3">Functional genomics of stromal cells in chronic inflammatory diseases</h1>

<p><b>Kamil Slowikowski</b>, Kevin Wei, Michael B. Brenner, Soumya Raychaudhuri</p>

<p><i>Current Opinion in Rheumatology</i>, 2018. DOI: <a rel="noopener" target="_blank" href="https://doi.org/10.1097/BOR.0000000000000455">10.1097/BOR.0000000000000455</a></p>

<p>
<a class="f5 fw5 link br-pill hvr-shadow ba b--black-10 ph3 pv2 mb2 dib near-black" rel="noopener" target="_blank" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5890939/pdf/nihms955133.pdf"><i class="far fa-file-pdf"></i> PDF</a>
<a class="f5 fw5 link br-pill hvr-shadow ba b--black-10 ph3 pv2 mb2 dib near-black" rel="noopener" target="_blank" href="https://www.ncbi.nlm.nih.gov/pubmed/28984647"><i class="fas fa-book-open"></i> PubMed</a>
</p>

<div class="figure mw6 center"><img src="/images/slowikowski2018.png" alt="Figure 1" data-zoomable></img></div>

<h2>Abstract</h2>

<p>PURPOSE OF REVIEW: Stroma is a broad term referring to the connective tissue
matrix in which other cells reside. It is composed of diverse cell types with
functions such as extracellular matrix maintenance, blood and lymph vessel
development, and effector cell recruitment. The tissue microenvironment is
determined by the molecular characteristics and relative abundances of
different stromal cells such as fibroblasts, endothelial cells, pericytes, and
mesenchymal precursor cells. Stromal cell heterogeneity is explained by
embryonic developmental lineage, stages of differentiation to other cell types,
and activation states. Interaction between immune and stromal cell types is
critical to wound healing, cancer, and a wide range of inflammatory diseases.
Here, we review recent studies of inflammatory diseases that use functional
genomics and single-cell technologies to identify and characterize stromal cell
types associated with pathogenesis.</p>

<p>RECENT FINDINGS: High dimensional strategies using mRNA sequencing, mass
cytometry, and fluorescence activated cell-sorting with fresh primary tissue
samples are producing detailed views of what is happening in diseased tissue in
rheumatoid arthritis, inflammatory bowel disease, and cancer. Fibroblasts
positive for CD90 (Thy-1) are enriched in the synovium of rheumatoid arthritis
patients, and also in prostate cancer tumors. Single-cell RNA-seq studies will
lead to more discoveries about the stroma in the near future.</p>

<p>SUMMARY: Stromal cells form the microenvironment of inflamed and diseased
tissues. Functional genomics is producing an increasingly detailed view of
subsets of stromal cells with pathogenic functions in rheumatic diseases and
cancer. Future genomics studies will discover disease mechanisms by perturbing
molecular pathways with chemokines and therapies known to affect patient
outcomes. Functional genomics studies with large sample sizes of patient
tissues will identify patient subsets with different disease phenotypes or
treatment responses.</p>

<!--<div class="center"><img src="/images/pdf-thumbs/slowikowski2018-montage.jpg" alt="Montage"></img></div>-->

</div>

<div id="2" class="bt b--black-10 pt5 mt5">

<h1 class="mt0 f3">Functionally distinct disease-associated fibroblast subsets in rheumatoid arthritis</h1>

<p>Fumitaka Mizoguchi&#42;, <b>Kamil Slowikowski</b>&#42;, Kevin Wei, Jennifer L. Marshall,
Deepak A. Rao, Sook Kyung Chang, Hung N. Nguyen, Erika H. Noss, Jason D.
Turner, Brandon E. Earp, Philip E. Blazar, John Wright, Barry P. Simmons, Laura
T. Donlin, George D. Kalliolias, Susan M. Goodman, Vivian P. Bykerk, Lionel B.
Ivashkiv, James A. Lederer, Nir Hacohen, Peter A. Nigrovic, Andrew Filer,
Christopher D. Buckley, Soumya Raychaudhuri &amp; Michael B. Brenner</p>

<p><i>Nature Communications</i>, 2018. DOI: <a rel="noopener" target="_blank" href="https://doi.org/10.1038/s41467-018-02892-y">10.1038/s41467-018-02892-y</a></p>

<p>
<a class="f5 fw5 link br-pill ba b--black-10 hvr-shadow ph3 pv2 mb2 dib near-black" rel="noopener" target="_blank" href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE109450"><i class="fas fa-box-open"></i> Data</a>
<a class="f5 fw5 link br-pill ba b--black-10 hvr-shadow ph3 pv2 mb2 dib near-black" rel="noopener" target="_blank" href="https://www.nature.com/articles/s41467-018-02892-y.pdf"><i class="far fa-file-pdf"></i> PDF</a>
<a class="f5 fw5 link br-pill ba b--black-10 hvr-shadow ph3 pv2 mb2 dib near-black" rel="noopener" target="_blank" href="https://www.ncbi.nlm.nih.gov/pubmed/29476097"><i class="fas fa-book-open"></i> PubMed</a>
</p>

<div class="figure mw6 center"><img src="/images/mizoguchi2018.jpg" alt="Figure 2" data-zoomable></img></div>

<h2>Abstract</h3>

<p>Fibroblasts regulate tissue homeostasis, coordinate inflammatory responses,
and mediate tissue damage. In rheumatoid arthritis (RA), synovial fibroblasts
maintain chronic inflammation which leads to joint destruction. Little is
known about fibroblast heterogeneity or if aberrations in fibroblast subsets
relate to pathology. Here, we show functional and transcriptional differences
between fibroblast subsets from human synovial tissues using bulk
transcriptomics of targeted subpopulations and single-cell transcriptomics.
We identify seven fibroblast subsets with distinct surface protein
phenotypes, and collapse them into three subsets by integrating
transcriptomic data. One fibroblast subset, characterized by the expression
of proteins podoplanin, THY1 membrane glycoprotein and cadherin-11, but
lacking CD34, is threefold expanded in patients with RA relative to patients
with osteoarthritis. These fibroblasts localize to the perivascular zone in
inflamed synovium, secrete proinflammatory cytokines, are proliferative, and
have an in vitro phenotype characteristic of invasive cells. Our strategy may
be used as a template to identify pathogenic stromal cellular subsets in
other complex diseases.</p>

<!--<div class="center"><img src="/images/pdf-thumbs/mizoguchi2018-montage.jpg" alt="Montage"></img></div>-->

</div>

<div id="1" class="bt b--black-10 pt5 mt5">

<h1 class="mt0 f3">SNPSEA: an algorithm to identify cell types, tissues, and pathways affected by risk loci</h1>

<p><strong>Kamil Slowikowski</strong>, Xinli Hu, Soumya Raychaudhuri</p>

<p><i>Bioinformatics</i>, 2014. DOI: <a rel="noopener" target="_blank" href="https://doi.org/10.1093/bioinformatics/btu326">10.1093/bioinformatics/btu326</a></p>

<p>
<a class="f5 fw5 link br-pill ba b--black-10 hvr-shadow ph3 pv2 mb2 dib near-black" rel="noopener" target="_blank" href="https://doi.org/10.6084/m9.figshare.871430.v4"><i class="fas fa-box-open"></i> Data</a>
<a class="f5 fw5 link br-pill ba b--black-10 hvr-shadow ph3 pv2 mb2 dib near-black" rel="noopener" target="_blank" href="https://github.com/slowkow/snpsea"><i class="fab fa-github"></i> GitHub</a>
<a class="f5 fw5 link br-pill ba b--black-10 hvr-shadow ph3 pv2 mb2 dib near-black" rel="noopener" target="_blank" href="https://academic.oup.com/bioinformatics/article-pdf/30/17/2496/17148439/btu326.pdf"><i class="far fa-file-pdf"></i> PDF</a>
<a class="f5 fw5 link br-pill ba b--black-10 hvr-shadow ph3 pv2 mb2 dib near-black" rel="noopener" target="_blank" href="https://www.ncbi.nlm.nih.gov/pubmed/24813542"><i class="fas fa-book-open"></i> PubMed</a>
</p>

<div class="figure mw6 center"><img src="/images/slowikowski2014.png" alt="Figure 1" data-zoomable></img></div>

<h2>Abstract</h3>

<p>We created a fast, robust and general C++ implementation of a
single-nucleotide polymorphism (SNP) set enrichment algorithm to identify cell
types, tissues and pathways affected by risk loci. It tests trait-associated
genomic loci for enrichment of specificity to conditions (cell types, tissues
and pathways). We use a non-parametric statistical approach to compute
empirical P-values by comparison with null SNP sets. As a proof of concept, we
present novel applications of our method to four sets of genome-wide
significant SNPs associated with red blood cell count, multiple sclerosis,
celiac disease and HDL cholesterol.</p>

<!--<div class="center"><img src="/images/pdf-thumbs/slowikowski2014-montage.jpg" alt="Montage"></img></div>-->

</div>
    
